### ISSUE TITLE
DeepSpeech won't train Common Voice data

### ISSUE THREAD

**Issue by raspi-three** [mozilla/DeepSpeech#1812]
For support and discussions, please use our [Discourse forums](https://discourse.mozilla.org/c/deep-speech).

If you've found a bug, or have a feature request, then please create an issue with the following information:

- **Have I written custom code (as opposed to running examples on an unmodified clone of the repository)**:
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
- **TensorFlow installed from (our builds, or upstream TensorFlow)**:
- **TensorFlow version (use command below)**:
- **Python version**: 
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**:
- **GPU model and memory**:
- **Exact command to reproduce**:

You can obtain the TensorFlow version with

```bash
python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"
```

Please describe the problem clearly. Be sure to convey here why it's a bug or a feature request.

Include any logs or source code that would be helpful to diagnose the problem. For larger logs, link to a Gist, not a screenshot. If including tracebacks, please include the full traceback. Try to provide a reproducible test case.

**Comment by lissyx** [comment]
> So why does this happen?

Isn't it processing files, thus using system RAM and running out of it? How much swap do you have?

**Comment by lissyx** [comment]
>  Swap is 2.0GB

So 10GB in total. We should be running 8 processes according to https://github.com/mozilla/DeepSpeech/blob/master/bin/import_cv.py#L95 and your CPU. I'm surprised it's using that much.

How much memory is really available?

**Comment by raspi-three** [comment]
HTOP via ssh shows Mem: 7.75G and Swp: 2.0G available. Yes, 8 processes are running. The system is running in the text mode. I have no idea were to go from here.

LOL I got deepspeech running on my raspberry pi with no issues!

The most significant use of memory use prior to start of training is snapd daemon which is using 0.3% 
Any suggestions?

**Comment by lissyx** [comment]
>  Any suggestions?

Check which processes are using the memory overtime? I'm wondering if you're just low in resources or we have something leaking ...

Also try more swap, just in case?

**Comment by raspi-three** [comment]
I increased the Swap size to 5GB. That appears to fix one issue. The preprocessing (valid and dev files) did complete using only RAM/Swap Memory. When it finished training did revert to the GPU. Since I only have 8GB of RAM, i will upgrade to 16GB (max for this machine). That should provide overall better performance during preprocessing. Much better than the use of a swap file.

However, I still think there is a issue with preprocessing, at least in my case. Here is what I observed with HTOP via SSH. The training machine was booted in the console mode.

After 30 minutes: Mem: 5GB/7.76GB  Swp: 0/5.0GB
After 40 minutes: Mem: 6.77GB/7.76GB  Swp:268KB/5.0GB
After 60 minutes: Mem: 7.57GB/7.76GB  Swp: 285MB/5.0GB
After 90 minutes: Mem: 7.57GB/7.76GB  Swp: 1.12GB\5.0GB
After 120 minutes: Mem: 7.59GB\7.76GB  Swp: 2.14GB\5.0GB
Within a few minutes later the preprocessing displayed Done and Training began.
This is where it got interesting.
After 150 minutes: Mem: 7.54GB\7.76GB  Swp: 4.46GB\5.0GB
After 7 hours: Mem: 7.56GB\7.76GB  Swp: 4.74GB\5.0GB

The next morning, training was still being done. The GPU processing observed in NVTOP via SSH looked normal and processing was being done as recorded in the checkpoint directory but RAM Mem was still indicating a continual 91.6% usage. If at this point, the only processing that should be happening is file retrieval for training and should not be that high or continuous. HTOP is showing 20-25 processes still in what appears to be in the preprocessing stage even though it displayed Done in the terminal window.

Your thoughts?

**Comment by lissyx** [comment]
> The next morning, training was still being done.

That's not clear to me. It seems it's still pre-processing audio files, not yet training. Now, as to why it takes that much memory, I'm a bit surprised, but we only run that on systems that have much more than what you have.

Honestly, training a big amount of audio data on a GTX1070 is going to take a lot of time itself, and preprocessing being slow will be the least of your concerns, at some point.

Still, preprocessing is mostly python code callling sox, I don't get why it would feel that much of memory. It might help if you could track better memory usage:
 - which processes are growing ?
 - is it really used memory or file cache ?

**Comment by raspi-three** [comment]
Just an update. After I upgraded my RAM memory to 16GB, everything is running fine. System is using 12.1GB and zero of 5GB in Swap. There are numerous files now in the checkpoint folder.

One quick question. Is there anyway to detect the percentage of training completed?

**Comment by kdavis-mozilla** [comment]
I assume you see the log, which indicates which epoch is running, and you know how many epochs are to be run, I assume. So, is that enough?

**Comment by bjtommychen** [comment]
> HTOP via ssh shows Mem: 7.75G and Swp: 2.0G available. Yes, 8 processes are running. The system is running in the text mode. I have no idea were to go from here.
> 
> LOL I got deepspeech running on my raspberry pi with no issues!
> 
@raspi-three  Hi, would you tell me which Pi board you use ? and how is the deepspeech inference speed? thanks.

**Comment by raspi-three** [comment]
@bjtommychen I use raspi 3 B+ with 1TB external drive. The speed is very good but not the accuracy.

**Comment by kdavis-mozilla** [comment]
@raspi-three Are you using our release model or a model of your own making?

**Comment by raspi-three** [comment]
I am using deepspeech 0.4.1 on ds-server. The pi is running again that server right now.

**Comment by raspi-three** [comment]
@kdavis-mozilla I can't seem to find the default location of the deepspeech log files. I assume it would be in one of the util files but I don't see it.

**Comment by kdavis-mozilla** [comment]
@raspi-three So you're using the deepspeech 0.4.1 code and the English model released with deepspeech 0.4.1?

**Comment by raspi-three** [comment]
yes that is correct

**Comment by kdavis-mozilla** [comment]
@raspi-three Why were you mentioning "Is there anyway to detect the percentage of training completed?" This sounds like you've trained your own model.

**Comment by raspi-three** [comment]
I was attempting to train common voice dataset. now I'm back to deepspeech
because I want to add on.

**Comment by kdavis-mozilla** [comment]
@raspi-three At what resolution and format are you capturing audio?

**Comment by raspi-three** [comment]
Captured with Blue Snowball mic at 44100 rate then used both numpy or audioop to convert to 16 bit 16K rate save to a wave file and passed to the deepspeech-server.

DeepSpeech train is still processing the CV files but I attempting to determine its effeciency. I still have not found "the log, which indicates which epoch is running".

**Comment by kdavis-mozilla** [comment]
If possible capture at 16 bit 16K mono.

Also, audioop does not convert to 16 bit 16K mono correctly, see for example this [issue](https://github.com/mozilla/DeepSpeech/issues/1726).

What is print out when you train?

**Comment by raspi-three** [comment]
I never found the means to capture directly from a microphone at 16/16. I shopped everywhere but all mics I found were either 44100 or 48000.

Thank for the issue on audioop.

The training is still in process, so I don't have a result yet. That is why I looking for a way to determine where deepspeech is in its training process.

**Comment by raspi-three** [comment]
My training machine has no monitor and boots in the console mode. So ssh to the machine I don't see any logging or errors. I since added a monitor and all is now functioning well.
